WordPress/Classes/ViewRelated/Post/PrivateSiteURLProtocol.m

```diff
@@ -114,51 +121,114 @@ - (void)startLoading
 {
     NSMutableURLRequest *mRequest = [self.request mutableCopy];
     [mRequest addValue:[NSString stringWithFormat:@"Bearer %@", [[self class] bearerToken]] forHTTPHeaderField:@"Authorization"];
-    NSURLSessionConfiguration *sessionConfiguration = [NSURLSessionConfiguration defaultSessionConfiguration];
-    self.session = [NSURLSession sessionWithConfiguration:sessionConfiguration delegate:self delegateQueue:nil];
-    self.sessionTask = [self.session dataTaskWithRequest:mRequest];
-    [self.sessionTask resume];
+    self.sessionTask = [[PrivateSiteURLProtocolSession sharedInstance] createSessionTaskForRequest:mRequest forProtocol:self];
+
 }
 
 - (void)stopLoading
 {
-    [self.sessionTask cancel];
-    [self.session invalidateAndCancel];
+    [[PrivateSiteURLProtocolSession sharedInstance] stopSessionTask:self.sessionTask];
     self.sessionTask = nil;
-    self.session = nil;
+}
+
+@end
+
+@interface PrivateSiteURLProtocolSession()
+
+@property (nonatomic, strong) NSMutableDictionary *taskToProtocolMapping;
+@property (nonatomic, strong) NSURLSession *session;
+
+@end
+
+@implementation PrivateSiteURLProtocolSession
+
++ (instancetype) sharedInstance
+{
+    static id _sharedInstance = nil;
+    static dispatch_once_t _onceToken;
+    dispatch_once(&_onceToken, ^{
+        _sharedInstance = [[self alloc] init];
+    });
+
+    return _sharedInstance;
+}
+
+- (instancetype)init
+{
+    self = [super init];
+    if (self) {
+        _taskToProtocolMapping = [NSMutableDictionary dictionary];
+    }
+    return self;
+}
+
+- (NSURLSession *)session
+{
+    if (_session == nil) {
+        NSURLSessionConfiguration *sessionConfiguration = [NSURLSessionConfiguration defaultSessionConfiguration];
+        _session = [NSURLSession sessionWithConfiguration:sessionConfiguration delegate:self delegateQueue:nil];
```

Comment:
This is the delegateQueue for callbacks from the session, we could get some
performance benefits but we could also run on memory issues by decoding
multiple images at the same time, even with the MaxConcurrentOperation set.

On Mon, Jul 18, 2016 at 5:57 PM, Eric <notifications@github.com> wrote:

> In WordPress/Classes/ViewRelated/Post/PrivateSiteURLProtocol.m
> <https://github.com/wordpress-mobile/WordPress-iOS/pull/5718#discussion_r71185436>
> :
>
> > +}
> > +
> > +- (instancetype)init
> > +{
> > +    self = [super init];
> > +    if (self) {
> > +        _taskToProtocolMapping = [NSMutableDictionary dictionary];
> > +    }
> > +    return self;
> > +}
> > +
> > +- (NSURLSession *)session
> > +{
> > +    if (_session == nil) {
> > +        NSURLSessionConfiguration *sessionConfiguration = [NSURLSessionConfiguration defaultSessionConfiguration];
> > +        _session = [NSURLSession sessionWithConfiguration:sessionConfiguration delegate:self delegateQueue:nil];
>
> Question: when passing nil to to delegateQueue the docs say a serial
> queue is used. Would it be better to either create a new NSOperationQueue
> or perhaps edit the one that's created and set maxConcurrentOperationCount
> to NSOperationQueueDefaultMaxConcurrentOperationCount?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/wordpress-mobile/WordPress-iOS/pull/5718/files/1fa7f22c2415afe7f5b7c0d2df94354aa7f58649#r71185436>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnxUQBf1DQXpKBwNr2AGsN211Nhnzv1ks5qW7BjgaJpZM4JMkkf>
> .
>